### PR TITLE
[Dynamo] Support graph breaks in checkpointed functions

### LINKED ISSUE #139989: Support AC with graph break
### 🐛 Describe the bug

As mentioned in this [blog](https://dev-discuss.pytorch.org/t/higher-order-operators-2023-10/1565), HigherOrderOperator does not support graph break inside the input/output function, because Dynamo cannot determine whether the operations inside the function are "safe". However, in most cases, graph break within the function is safe and acceptable. Is there a way to make Dynamo support these cases without falling back to eager mode? For example, this scenario frequently occurs in LLM models:

```
import torch
import torch.nn as nn
import flash_attn


class Attention(nn.Module):
    def __init__(self):
        super().__init__()
        self.q_proj = nn.Linear(4096, 4096)
        self.k_proj = nn.Linear(4096, 4096)
        self.v_proj = nn.Linear(4096, 4096)
        self.o_proj = nn.Linear(4096, 4096)
    
    def forward(self, x):
        def attn_forward(x):
            q = self.q_proj(x)
            k = self.k_proj(x)
            v = self.v_proj(x)
            q = q.view(x.shape[0], x.shape[1], 64, 64).transpose(1, 2)
            k = k.view(x.shape[0], x.shape[1], 64, 64).transpose(1, 2)
            v = v.view(x.shape[0], x.shape[1], 64, 64).transpose(1, 2)
            return flash_attn.flash_attn_func(q, k, v)
        return torch.utils.checkpoint.checkpoint(attn_forward, x, use_reentrant=False)

model = Attention()
model.cuda()

def custom_backand(gm, example_inputs):
    print(gm)
    return gm

opt_model = torch.compile(model, backend=custom_backand)

x = torch.randn(1, 32, 4096).cuda()

with torch.autocast("cuda", dtype=torch.bfloat16):
    out = opt_model(x)
```

In LLM models, flash attention is often used along with activation checkpointing, which can cause the entire attention or decoder layer execution to fall back to eager mode.

Is there a way to make Dynamo support graph breaking within HigherOrderOperator, such as providing some APIs to mark these functions as safe?

I know that we can use PyTorch's SDPA (Scaled Dot-Product Attention) op as a replacement, but our users typically hard-code the use of flash attention in their code. I hope there is a way to achieve this without requiring significant changes to the user's code.

### Error logs

_No response_

### Minified repro

_No response_

### Versions

PyTorch version: 2.5.0a0+git24dee99
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-2ubuntu1~20.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.31

Python version: 3.10.13 (main, Aug 25 2023, 13:20:03) [GCC 9.4.0] (64-bit runtime)

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.5.0a0+git417a076
[pip3] torchvision==0.18.0

cc @ezyang @gchanan @kadeng @msaroufim @soulitzer @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @amjames @Lucaskabela @jataylo @ydwu4 @bdhirsh @bobrenjc93 @aorenste @zou3519 @chenyang78 @yf225

### PR DESCRIPTION
**Pull request by guan404ming**
## Related Issue

Fixes #139989.

## Why

- A graph break inside a checkpointed function made Dynamo skip the entire frame.
- The outer forward and the checkpointed function both ran eager, losing all speedups.
- Implements the fallback design @zou3519 proposed on the issue: turn `compile(ac(f))` with a graph break in `f` into `ac(compile(f))`, which already works.

## How

- Fall back to an eager checkpoint call that re-enters Dynamo on the checkpointed function.
- Run forward and backward recompute through the same compiled code, keeping recompute consistent.
- Keep raw `tag_activation_checkpoint` usage and `fullgraph=True` behavior unchanged.

Authored with an AI assistant. Reviewed by me.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by anijain2305 — CHANGES_REQUESTED** [review]
As noted in the https://docs.google.com/document/d/1nsO52Q74VXxrwOdGxaNEVEedy-6Jra-FyYjWQxnVVxE/edit?tab=t.6wl6v2b8z2mt , the reordering of `compile(ac(f))` to `ac(compile(f))` does not work for selective activation checkpointing.

> AI-generated review (Codex):
>
> Could we verify that the SAC policy is actually honored in this fallback? `test_graph_break_selective_checkpoint` only checks outputs and gradients with `aot_eager`; it does not show that `MUST_SAVE` prevents the `mm` from being recomputed. With Inductor, the graph-break-separated compiled regions are opaque to SAC by default because `wrap_inductor_compiled_regions` is disabled, so the policy may silently degrade to recomputing an entire region. Please add an execution-count test covering `MUST_SAVE` and `PREFER_RECOMPUTE` with the production backend, or explicitly reject this combination.

**Comment by guan404ming** [comment]
Good catch, thanks. With inductor the mm is fused into the compiled region, so MUST_SAVE silently degrades to full recompute. Fixed by keeping SAC  (non-default context_fn) eager in the fallback. Added an aten.mm execution-count test for MUST_SAVE vs PREFER_RECOMPUTE on aot_eager and inductor. Pushed in 9e50f20.

**Comment by anijain2305** [comment]
> Good catch, thanks. With inductor the mm is fused into the compiled region, so MUST_SAVE silently degrades to full recompute. Fixed by keeping SAC (non-default context_fn) eager in the fallback. Added an aten.mm execution-count test for MUST_SAVE vs PREFER_RECOMPUTE on aot_eager and inductor. Pushed in [9e50f20](https://github.com/pytorch/pytorch/commit/9e50f20d25f9f8ddd25b8c02bc0356316190dba0).

This is too much fragmentation now.
* compile traces no-SAC-checkpointed function if there is a graph break
* compile does not trace SAC-checkpointed if there is a graph break

This will cause a lot of confusion.

**Comment by guan404ming** [comment]
Agreed, the split is kind of confusing for users. 

Here is my thought: AC and SAC both take the same `checkpoint(compile(f))` fallback. For SAC, since the policy cannot see ops inside a compiled region, warn that the policy is not honored across the graph break instead of silently going eager. Keeping one rule and no hidden perf cliff.

Making the policy work across graph breaks (route `context_fn` into each compiled subgraph, handle stateful policies) is a little bit larger. I would do it as a follow-up. Happy to make it a hard error instead of a warning if you prefer.

**Comment by anijain2305** [comment]
> Agreed, the split is kind of confusing for users.
> 
> Here is my thought: AC and SAC both take the same `checkpoint(compile(f))` fallback. For SAC, since the policy cannot see ops inside a compiled region, warn that the policy is not honored across the graph break instead of silently going eager. Keeping one rule and no hidden perf cliff.
> 
> Making the policy work across graph breaks (route `context_fn` into each compiled subgraph, handle stateful policies) is a little bit larger. I would do it as a follow-up. Happy to make it a hard error instead of a warning if you prefer.

So, I have not given much thought on how to make this work with SAC. It seems pretty hard on the first glance.

In situations like these, the usual stance of a maintainer is to first see the full design and then implement. I worry that without full design, we will lose interest in solving SAC, and we will live in a fragmented state for future.

**Comment by guan404ming** [comment]
Thanks for the feedback, agreed the design should come first. I wrote it up here: https://docs.google.com/document/d/1vFzL99cntyay1ti-ukSjwfg_NOTziY8XPUpROK5upMo/edit?tab=t.0#heading=h.56fivbdsf9z9. Feel free to comment or correct me if there is anything issue, thanks!

Here is the short version:
1. Drop the eager `checkpoint()` fallback.
2. While `f` runs with Dynamo re-enabled, every compiled frame (including resume frames after the break) becomes its own `tag_activation_checkpoint` HOP
with the same `context_fn`.
3. Each region goes through the existing trace-time policy tagging and partitioner. Policy honored per aten op. One mechanism for AC and SAC.
4. Ops executed eagerly at the break are saved, not recomputed. Extra memory is bounded and explicit.

One small question: This has the same shape as a HOP body breaking natively under nested graph breaks. Is there a timeline for HOP support there? If close, I would rather build on it than add thread-local checkpoint state.

Happy to implement once you are fine with the approach.
